File: lib/compose.js
'use strict';

var fs = require('fs');
var yaml = require('js-yaml');

module.exports = {
  addService: function(name, config, override) {
    var configFile = readBase();

    configFile.services[name] = config || null;

    writeBase(configFile);

    if (override) {
      var overrideFile = readOverride();

      overrideFile.services[name] = override;

      writeOverride(overrideFile);
    }
  },

  removeService: function(name) {
    var configFile = readBase();
    var overrideFile = readOverride();

    delete configFile.services[name];
    delete overrideFile.services[name];

    writeBase(configFile);
    writeOverride(overrideFile);
  }
};

function readBase() {
  return normalize(yaml.safeLoad(fs.readFileSync('docker-compose.yml', 'utf-8')));
}

function readOverride() {
  return normalize(yaml.safeLoad(fs.readFileSync('docker-compose.override.yml', 'utf-8')));
}

function writeBase(data) {
  fs.writeFileSync('docker-compose.yml', yaml.safeDump(data));
}

function writeOverride(data) {
  fs.writeFileSync('docker-compose.override.yml', yaml.safeDump(data));
}

function normalize(data) {
  data.services = data.services || {};
  return data;
}

File: lib/frigate/compose.rb
require 'yaml'

module Frigate
  class Compose
    def add_service(name, config, override = nil)
      config_file = read_base
      config_file['services'][name] = config

      write_base config_file

      if override
        override_file = read_override
        override_file['services'][name] = override

        write_override override_file
      end
    end

    def remove_service(name)
      config_file = read_base
      override_file = read_override

      config_file['services'].delete(name)
      override_file['services'].delete(name)

      write_base config_file
      write_override override_file
    end

    private

    def read_base
      normalize(YAML.load(File.read('docker-compose.yml')))
    end

    def read_override
      normalize(YAML.load(File.read('docker-compose.override.yml')))
    end

    def write_base(data)
      File.write('docker-compose.yml', data.to_yaml)
    end

    def write_override(data)
      File.write('docker-compose.override.yml', data.to_yaml)
    end

    def normalize(data)
      data['services'] ||= {}
      data
    end
  end
end
